### DISCUSSION TITLE
open rofi from another rofi with -normal-window

### DISCUSSION THREAD

**Discussion by numberokta** [davatorium/rofi#1883] [Ideas/Feature Requests]
Hi, let me explain what I'm asking. I open Rofi-1 with -normal-window in script mode because I want to open another rofi from it. I unseat ROFI_OUTSIDE and eventually open rofi-2  with -normal-window like two separate window apps.
When I open Rofi-2, Rofi-1 will freeze, but if I click on some button in Rofi-1, it receives a click action and holds it until Rofi-2 is closed, and then it will do whatever that button does. If I click that button 10 times, after closing Rofi-2, it does the function related to that button 10 times.
I would like to know if there is any way to freeze Rofi-1 completely that it does nothing when there is another app open from it ?

**Comment by DaveDavenport** [comment]
No idea, this is not something tested/supported.

**Reply by numberokta** [reply]
Can you test it ?
It's easy, just create an rofi in script mode with -normal-window and unset ROFI_OUTSIDE and then open another app.

**Reply by numberokta** [reply]
I'm done it by making some changes in the source code.

**Reply by DaveDavenport** [reply]
Can you share a patch?

**Reply by numberokta** [reply]
See [this](https://github.com/numberokta/rofi)
Create a Rofi in **script mode** and in whatever **state** you want you can do this to freeze Rofi :
```sh
freeze=/path/freeze
touch "$freeze"
echo -e "\0freeze\x1f$freeze"
# open another Rofi or any app (it freezes Rofi)
coproc(sleep 0.5;rm "$freeze") # this unfreezes Rofi
```

**Reply by DaveDavenport** [reply]
Isn't this already possible via a signal?  sending a SIGSTOP and SIGCONT to continue?

**Reply by numberokta** [reply]
No. Sending SIGSTOP signal just freeze Rofi, but it receives key bindings and hold them until receives SIGCONT signal and then do whatever key bindings do.

**Reply by DaveDavenport** [reply]
I think the above solution is not right, if we would want to fix this, we should probably release the keyboard/mouse grab on focus out event instead. Relying on magic 'lock' files does not sound right.

(or maybe not take pointer/keyboard in normal window mode?, not sure what it currently does)

^^ if this is the actual issue.

**Reply by DaveDavenport** [reply]
Do you have a script to reproduce your problem, this works fine for me?

```bash
#!/usr/bin/env bash
unset ROFI_OUTSIDE

if [ x"$@" == x"quit" ]
then
  exit;
fi
if [ $ROFI_RETV != 0 ]
then
  coproc(rofi -show drun -normal-window)
  echo "launch rofi"
  echo "quit"
else
  echo "launch rofi"
  echo "quit"
fi
```

**Reply by numberokta** [reply]
Test this:
```sh
#!/usr/bin/env bash
unset ROFI_OUTSIDE

if [ x"$@" == x"quit" ]
then
  exit;
fi
if [ $ROFI_RETV != 0 ]
then
  # Echo chosen one, from child rofi to parent rofi
  echo -e "test-1\ntest-2\ntest-3" | rofi -dmenu -normal-window
  echo "launch rofi"
  echo "quit"
else
  echo "launch rofi"
  echo "quit"
fi
```
When dmenu rofi is open, parent rofi is frozen and this is ok. Now click on quit in the parent rofi. You see nothing happen. Then close demnu rofi and now you can see quit is selected.
It holds all functions related to parent rofi and when child process was closed, it does those functions.

**Reply by DaveDavenport** [reply]
Yes, what you are doing above is terribly wrong.    scripts are blocking, so you are blocking rofi and sending it commands (X messages will queue up) and then once unblock it obviously starts handling this.

Do not nest rofi, this is also why rofi tries to stop you from doing this.

So we need to either make script mode non-blocking (is on the todo) or be more forcefull then a environment setting to stop you from doing this.

**Reply by numberokta** [reply]
I think it's not wrong. Complex apps can be made in this way. I create so many apps with my patched rofi. Actually, I created an entire linux distribution based on my patched rofi.
Can you tell me what problem could be made this way and what could be gone wrong ?

**Reply by DaveDavenport** [reply]
> I think it's not wrong. 

It is a hack and not a proper fix.  You have a rofi instance hanging not responding to the system.
The script should run async from rofi, so rofi  can keep responding to input from the system.
As I mentioned before this is on the todo (an implementation of this is already used in dmenu/recursivebrowser that needs to be ported over).

>  create so many apps with my patched rofi. Actually, I created an entire linux distribution based on my patched rofi.

That does not make the "fix" right.

I would not like people using 'rofi' with a hack like this,  you have one instance of rofi (or multiple that are nested) that are hanging .
If you distribute this to others, please make sure people are aware they are using a non-stock broken version of rofi, not rofi.

**Reply by numberokta** [reply]
You are right. It's a hack for now and has some issues. I have some other ideas that i must test them first.

**Reply by numberokta** [reply]
I have idea to update rofi script mode with `SIGUSR1` signal.
I add this codes to `/source/view.c`
```c
void rofi_view_signal(int signum) {
  RofiViewState *state = rofi_view_get_active();
  // execute mode.sh with ROFI_RETV=3
  // i don't know how
}
```
and add this code to `rofi_view_create` function
```c
signal(SIGUSR1, rofi_view_signal);
```
This way, we can use `coproc` to open another apps without blocking rofi and then update resaults from a file with sending `SIGUSR1` signal.

**Reply by DaveDavenport** [reply]
I think the right fix is to make the reading of the script output async (as mentioned a few times). 
This way it does not matter if the script blocks and it will show the output as it comes in.
This will solve several issues currently existing, instead of making hacks around it.

It is not trivial, but it is the right fix and not a hack.

**Reply by numberokta** [reply]
Can you explain to me how ?
